File: leetcode/cpp/P0673_NumberOfLIS.cpp
// https://leetcode.com/problems/number-of-longest-increasing-subsequence/

#include <algorithm>
#include <vector>

class Solution {
 public:
  int findNumberOfLIS(std::vector<int>& nums) {
    std::vector<int> lengths(nums.size(), 0);
    std::vector<int> counts(nums.size(), 1);

    int thelength = 0;

    for (int ii = 0; ii < nums.size(); ++ii) {
      for (int jj = 0; jj < ii; ++jj) {
        if (nums[jj] < nums[ii]) {
          if (lengths[jj] >= lengths[ii]) {
            lengths[ii] = lengths[jj] + 1;
            counts[ii] = counts[jj];
          } else if (lengths[jj] + 1 == lengths[ii]) {
            counts[ii] += counts[jj];
          }
        }
      }
    }

    int maxlength = 0;
    int cnt = 0;
    for (int ii = 0; ii < nums.size(); ++ii) {
      maxlength = std::max(lengths[ii], maxlength);
    }
    // cout << maxlength << " ";
    for (int ii = 0; ii < nums.size(); ++ii) {
      if (lengths[ii] == maxlength) {
        cnt += counts[ii];
      }
    }

    return cnt;
  }
};

File: leetcode/python/P0673_NumberOfLIS.py
# https://leetcode.com/problems/number-of-longest-increasing-subsequence/
from typing import List


class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        # len(nums) <= 2000 --> O(n^2) should work.
        # keep dp[i] = length of LIS up to i, inclusive.
        # for current i, take the max of previous values.

        # use Counter to get the count of max element in dp[i]
        # or just keep track of max value, i.e. length of LIS

        dp = [1] * len(nums)
        cnt = [0] * len(nums)

        max_len = 1

        cnt[0] = 1
        for i in range(1, len(nums)):
            for j in range(0, i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)
                    max_len = max(max_len, dp[i])
            # once all lengths are updated
            # count how many sequences of length dp[i]
            # end here.
            for j in range(0, i):
                if nums[i] > nums[j] and dp[j] == dp[i] - 1:
                    cnt[i] += cnt[j]
            cnt[i] = max(cnt[i], 1)

        res = 0
        for d, c in zip(dp, cnt):
            if d == max_len:
                res += max(1, c)

        return res


# TODO testcases
# [1,3,5,4,7]
# [2,2,2,2,2]
# [1,2,4,3,5,4,7,2]
# [1,3,2]
# [1,1,1,2,2,2,3,3,3]
